### PR TITLE
Added support for compute to `BinaryArray`

### LINKED ISSUE #345: Add support for binary compute
As [datafuselabs/datafuse:#1601](https://github.com/datafuselabs/datafuse/issues/1601), datafuse plans to remove `DataType::Utf8` and rename `DataType::Binary` to `DataType::String`.  The `DataType::String` is the logic datatype of `DataType::LargeBinary` in arrow.  Therefore, it is necessary to add binary support for the arraw compute. I will create pr for the issue.

### PR DESCRIPTION
**Pull request by zhyass**
This PR improves the support of compute to `BinaryArray`. Main additions:
* comparison
* sort
* contains
* like
* cast
* substring
*  min_max

Closes #345

### CODE REVIEW

**Comment by jorgecarleitao** [comment]
IMO this is ready to review (and merge ^_^)

**Comment by zhyass** [comment]
> IMO this is ready to review (and merge ^_^)

Work in progress, I will complete soon.

File: src/array/binary/mutable.rs [hunk 1]
```diff
@@ -148,6 +179,90 @@ impl<O: Offset, P: AsRef<[u8]>> FromIterator<Option<P>> for MutableBinaryArray<O
...
+        I: TrustedLen<Item = std::result::Result<Option<P>, E>>,
+    {
+        // soundness: I: TrustedLen
+        unsafe { Self::try_from_trusted_len_iter_unchecked(iterator) }
+    }
+
+    /// Creates a new [`MutableBinaryArray`] from a [`Iterator`] of `&[u8]`.
+    pub fn from_iter_values<T: AsRef<[u8]>, I: Iterator<Item = T>>(iterator: I) -> Self {
+        let (offsets, values) = values_iter(iterator);
+        // soundness: T: AsRef<[u8]>
```
Comment (jorgecarleitao) [inline src/array/binary/mutable.rs:261]:
```suggestion
```

File: src/compute/contains.rs [hunk 2]
```diff
@@ -96,6 +96,40 @@ where
     Ok(BooleanArray::from_data(DataType::Boolean, values, validity))
 }
 
+/// Checks if a [`GenericListArray`] contains a value in the [`BinaryArray`]
```
Comment (jorgecarleitao) [inline src/compute/contains.rs:99]:
```suggestion
/// Checks if a [`ListArray`] contains a value in the [`BinaryArray`]
```

**Review by jorgecarleitao — APPROVED** [review]
Amazing work, @zhyass! Went through this carefully and have only very tiny improvements, looks pretty solid, including tests.

**Pull request merged**